src/net_processing.cpp

```diff
@@ -3769,24 +3683,15 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
     }
 
     if (msg_type == NetMsgType::NOTFOUND) {
-        // Remove the NOTFOUND transactions from the peer
-        LOCK(cs_main);
-        CNodeState *state = State(pfrom.GetId());
         std::vector<CInv> vInv;
         vRecv >> vInv;
-        if (vInv.size() <= MAX_PEER_TX_IN_FLIGHT + MAX_BLOCKS_IN_TRANSIT_PER_PEER) {
+        if (vInv.size() <= MAX_PEER_TX_ANNOUNCEMENTS + MAX_BLOCKS_IN_TRANSIT_PER_PEER) {
```

Comment:
I'm not sure there is really a problem here, because:

* In step 2, that relay will take several minutes (we relay at most 35 tx invs per message, `INVENTORY_BROADCAST_MAX`, and around 1 inv message every 2s on outbound connections, so 17.5 tx/s), so announcements will come in gradually.
* In step 5, Bob will expire requests after 1 min anyway, even without NOTFOUNDs.

Reply:
> In step 2, that relay will take several minutes (we relay at most 35 tx invs per message, INVENTORY_BROADCAST_MAX, and around 1 inv message every 2s on outbound connections, so 17.5 tx/s), so announcements will come in gradually.

Yes but I'm wondering if you can bypass this by forcing to relay orphans with a lot of missing parents, thus the receiver marking them as announcements from the sender.

* Alice announces 17 txn to Bob
* Bob requires them all
* Alice replies with the 17 txn
* Bob receives the txn, all of them are orphans with 295 missing parents. Bob mark all parents as `AddTxAnnouncement` (src/net_processing.cpp, L2991) as coming from Alice
* Alice can't announce new transactions to Bob until her MAX_PEER_TX_ANNOUNCEMENTS is dried-up

I'll write a test to get certainty on behavior. Feel free to mark as resolved.

> In step 5, Bob will expire requests after 1 min anyway, even without NOTFOUNDs.

Right, but that's still 1min of potential tx-relay severance...

Reply:
> Yes but I'm wondering if you can bypass this by forcing to relay orphans with a lot of missing parents, thus the receiver marking them as announcements from the sender.

Mallory can't cause Alice to relay orphans to Bob - they're only relayed when they enter the mempool, which implies the parents are known.

> I'll write a test to get certainty on behavior. Feel free to mark as resolved.

Please do - though this sounds like something you'd need a functional test for, as it's not purely TxRequestTracker logic.

> Right, but that's still 1min of potential tx-relay severance...

Yeah, that would be a problem if an attacker can cause it with high probability, on many peers of a victim at once.

Reply:
> Mallory can't cause Alice to relay orphans to Bob - they're only relayed when they enter the mempool, which implies the parents are known

What is an orphan for Bob might not be an orphan for Alice if their mempools are divergent ? Mallory broadcast tx123 to Alice and tx123' to Bob, then send children from tx123 to Alice which will be relayed to Bob ? `fRejectedParents` checks parents, so Mallory may need to another layer of transaction, so it would be tx123 <- tx456 <- tx789. I think tx789 will be considered as a valid orphan by Bob ?

Reply:
Ok, so Mallory also knows the victim specifically, and is able to connect directly to them.

I believe you may be right that if Mallory has connections to both Alice and Bob, he may be able to induce a stream of orphans between them. If that stream is resolved slow enough (due to processing speed, bandwidth, ...), it may be possible to exceed the 5000 MAX_PEER_TX_ANNOUNCEMENTS limit. I think in this situation:
* This may be costly as the attacker has produce huge amounts of actually valid transactions to get Alice to relay them to Bob - transactions which may confirm.
* The limit would only be reached very temporarily, as new announcements get dropped once the limit is reached, and receiving a tx or NOTFOUND will decrease it again. So this attack would need to be maintained in a rapid fire to keep the tracked announcements at the bound.
* If the load comes very quickly, the affected transactions will primarily the attacker's transactions that we don't really care about. If the load doesn't come very quickly, it won't do much.
* This seems harder to pull off, and less effective, than simply mounting an invblock attack on Bob directly (race the network, announce honest transactions very quickly to Bob, and don't respond to requests).

Does that match your understanding?

Reply:
> This may be costly as the attacker has produce huge amounts of actually valid transactions to get Alice to relay them to Bob - transactions which may confirm.

Assuming a mass-connecting malicious node, I think you can mempool-partition from the rest of the honest network, Alice and Bob, by announcing yet-another different spend tx123''. Thus not having to pay for the malicious stream of orphans.

> The limit would only be reached very temporarily, as new announcements get dropped once the limit is reached, and receiving a tx or NOTFOUND will decrease it again. So this attack would need to be maintained in a rapid fire to keep the tracked announcements at the bound.

If you assume Mallory has a connection with Bob, can Mallory invblock on the malicious orphan stream by announcing it at the same time ? At least for an expiration period.

> In order to actually block Bob from receiving hones transactions, Mallory has to do this with a significant fraction of Bob's peers - not just Alice.

I was thinking the reverse, an attacker blocking Alice's txn from propagating. E.g in a scenario where the attacker expect Alice to broadcast a time-sensitive tx after reception of a given block. But agree an attacker has to repeat the blocking with every Alice's peers.

Overall, I agree it sounds hard to pull off, an attacker already has to observe all victim's tx-relay peers to effectively censor an expected broadcast. I think future p2p work like outbound tx-relay rotation should harden against this kind of attack scenario.